api/src/services/rapid-pro.js

```diff
@@ -0,0 +1,240 @@
+   * @returns {Promise<[stateUpdate]>} - array of state update objects.
+   */
+  send: (messages) => {
+    // get the credentials every call so changes can be made without restarting api
+    return getCredentials().then(credentials => {
+      let promiseChain = Promise.resolve([]);
+      messages.forEach(message => {
+        promiseChain = promiseChain.then((statusUpdates) => {
+          return sendMessage(credentials, message).then(result => {
+            if (result) {
+              statusUpdates.push(result);
+            }
+
+            return statusUpdates;
+          });
+        });
+      });
+
+      return promiseChain;
+    });
+  },
+
+  /**
+   * Polls RapidPro `messages` endpoint to check for state updates of outgoing messages that
+   * are in non-final states. Queries `medic-sms/gateway_messages_by_state` for a batch of messages that are in
+   * non-final states, `skip`ping rows that were already processed. Increases the global variable `skip` with the
+   * number of rows that were not updated (their states have not changed) from this batch.
+   * When there are no more rows to process, sets `skip` to 0 and will start over the queue on next iteration.
+   */
+  poll: () => {
+    // get the credentials every call so changes can be made without restarting api
+    return getCredentials()
+      .then(credentials => {
+        const viewOptions = { keys: _.uniq(nonFinalStates), limit: BATCH_SIZE, skip };
+        return db.medic
+          .query('medic-sms/gateway_messages_by_state', viewOptions)
+          .then(result => {
+            if (!result || !result.rows || !result.rows.length) {
+              skip = 0; // start from the beginning on the next poll call
+              return;
+            }
+
+            // avoid circular dependency issues, the messaging service also requires this file
+            // todo reviewer: I tried breaking up messaging, to expose the updateMessageTaskStates in a separate module,
+            // but it didn't turn out quite right. a "proper" refactor would have ended up moving more code around.
+            // Thoughts?
+            const messaging = require('./messaging');
+
+            const messages = result.rows.map(row => row.value);
+            return getRemoteStates(credentials, messages)
+              .then(statusUpdates => messaging.updateMessageTaskStates(statusUpdates))
+              .then(({ saved = 0 }={}) => {
+                // Only increase the skip with the number of messages that were *not updated*.
+                // The messages that were updated could have changed to be in a final state
+                // and will be excluded on the next query.
+                const numberOfRowsProcessed = result.rows.length - saved;
+                skip += numberOfRowsProcessed;
+              });
+          });
+      })
+      .catch(err => {
+        logger.error('Error while polling message states: %o', err);
+      });
+  },
```

Comment:
If I'm reading this correctly then the most statuses we can update is 25 per minute. If we have a project with more throughput than that it'll never catch up. I think we need to loop through all non-final messages with every poll call so we can go as fast as possible.

This will exacerbate another issue where `poll` may get called before the previous `poll` has completed which could cause conflicts or at least wasted effort. The naive solution would be to just have a `running` flag. A more nuanced approach would be to not schedule the next poll call until the first one returns.

Reply:
> I think we need to loop through all non-final messages with every poll call so we can go as fast as possible.

I actually implemented this in one iteration. Exactly with a `running` flag. I can rollback to that :) 
I'll have to remember why I went with the 25/minute, there was a reason but it eludes me now.

Reply:
I think the reason why I didn't want to keep recursively calling the messages endpoint was because i was worried of the rapidpro request/hour limit and what happens if the limit is hit.

Reply:
My biggest concern, I think, was if someone switched from another outgoing message service to rapid-pro, and they had a bunch of older messages that are in non-final states. 
I was wondering if we could add another property to the message with the outgoing message service that was used to generate the "gateway_ref" and have the new view index by that value and the state. 
But then I thought it was overkill. 

Now I'm unsure again. Please share your thoughts about this.